vllm/model_executor/layers/fused_moe/fused_moe.py

```
@@ -1160,6 +1146,101 @@ def flashinfer_fused_moe_blockscale_fp8_fake(
 )
 
 
+def flashinfer_fused_moe_per_tensor_scale_fp8(
+    routing_logits: torch.Tensor,
+    routing_bias: Optional[torch.Tensor],
+    hidden_states: torch.Tensor,
+    input_scale: torch.Tensor,
+    gemm1_weights: torch.Tensor,
+    gemm1_weights_scale: torch.Tensor,
+    activation_scale: torch.Tensor,
+    gemm2_weights: torch.Tensor,
+    gemm2_weights_scale: torch.Tensor,
+    num_experts: int,
+    top_k: int,
+    num_expert_group: Optional[int],
+    topk_group: Optional[int],
+    intermediate_size: int,
+    local_expert_offset: int,
+    local_num_experts: int,
+    use_routing_scales_on_input: bool,
+    routed_scaling_factor: float = 1.0,
+    routing_method_type: int = 3  # Llama4-styled routing method
```

I'm not a fan of defaulting this parameter if it is going to dictate model support. For instance in the current usage of this function this parameter isn't set, but there is no check that this is model needs llama 4 routing i.e. it would be silently incorrect for a Mixtral with the same quant

---

@amirkl94 Maybe let's remove the default value for `routing_method_type` and make this arg a required argument?

And from `llama4.py` we should pass this into fused_moe.py?

---

I think I can't check `custom_routing_function == Llama4MoE.custom_routing_function`, unless you meant in `llama4.py`?

Should I just make this parameter optional and pass it only from `llama4` and if it's not passed I'll default to the non-flashinfer implementation?

---

Yes this was what I meant. Obviously not optimal, but should be okay

---

@mgoin Currently, FlashInfer's per-tensor FP8 MoE only supports Llama4 routing mode, so I told @amirkl94 to assert if `layer.routing_method == Llama4MoE.custom_routing_function` is True. If it is not, an exception will be raised.

This is done such that in the future if anyone wants to use FlashInfer per-tensor FP8 MoE for another model, it will fail loudly telling the user why that is not supported. My philosophy is: a loud failure is better than a silent corruption.

Could you check if the current implementation is acceptable to you? Thanks!